### ISSUE TITLE
Similarity-based Malicious Prompt Detection

### ISSUE THREAD

**Issue by gventuri** [Sinaptik-AI/pandas-ai#456]
## Problem:
Currently, PandasAI lacks a robust mechanism to detect malicious prompts that instruct the model to ignore the previous text and generate harmful output. While pattern matching helps catch some of these prompts, it might not cover all variations.

## Solution:
I suggest implementing a similarity-based detection system using techniques like cosine similarity between prompts. This approach would involve maintaining a list of known malicious prompt patterns and leveraging NLP techniques to compare user-provided prompts against this list.

## Proposed Implementation:
-Create a list of known malicious prompt patterns, incorporating common variations of instructions to ignore previous text and generate malicious output.
- Use a suitable NLP library (e.g., sklearn, transformers) to vectorize prompts and calculate cosine similarity against the list of malicious patterns.
- Set a similarity threshold to determine when a prompt is potentially malicious. If the cosine similarity between the user's prompt and any of the malicious patterns exceeds the threshold, the system should raise a warning or prevent the execution of the prompt.

## Expected Benefits:
Improved detection of malicious prompts with varying wording and structure.
Enhanced user safety by preventing potentially harmful model outputs.
Reduced risk of inappropriate or harmful content generated by the AI.

## Challenges:
Ensuring a balance between sensitivity and specificity of the similarity-based detection.
Handling false positives and negatives by fine-tuning the system and considering user feedback.

### Motivation, pitch

Currently, we lack a mechanism to prevent this kind of attacks, making it possible prompt injections.

**Comment by fakabbir** [comment]
## Proposal Options

There are two possible ways to approach the problem. With its own pros and cons
**1. Inline Malicious Code Scanner
2. Available as an external package to wrap the code to scan.**

### With approach 1, 

the scanning in done within the code, it's a default option where the dependencies are installed with the library and does not require installation. 

The cons for this approach is that the library would get big in size as it has to install  nlp libraries and models to determine malicious nature of text. The high level usage would be as shown below
```
scanner = Scanner(config)
df = SmartDataframe(df, config={"llm": llm, "scanner": scanner})
```


### Where Approach 2,

 is a parallel development approach where we work on the malicious code detection task as a separate library and do the integration modification in pandasai. The advantage for this would be that the user who knows the queries or only want to explore for themselves and do not require the packages would not be troubled with large size of package and complexity. Further the definition of malicious might change based on use case as well. 

For someone trying to generate a workflow for cleaning data frame, deleting might  not a malicious code. These differences would create a wide scope that could dilute the core problem pandasai is solving.

## Advised Proposal:
Start with approach 2 to create an independent package which can be installed by providing custom variations during the installation step.

### High level changes

#### Installation step
Current : `pip install pandasai`
After modification
```
pip install pandasai[with_scanner]
```
The setup file would look like something this.
```
#setup.py file 
from setuptools import setup

setup(
    name='pandasai',
    install_requires=[
        'numpy', ....
    ],
    extras_require={
        'with_scanner': ['with_scanner'],
    },
)
```

#### Integration Step:
A separate module would be used to call the external library to get the results for us.
```
# pandasai/scanner/__init__.py
import with_scanner

def Scanner(config):
    pass
```

The code usage would be
```
from pandasai.scanner import Scanner

import pandas as pd
from pandasai import SmartDataframe

# Sample DataFrame
df = pd.DataFrame({
    "country": ["United States", "United Kingdom", "France", "Germany", "Italy", "Spain", "Canada", "Australia", "Japan", "China"],
    "gdp": [19294482071552, 2891615567872, 2411255037952, 3435817336832, 1745433788416, 1181205135360, 1607402389504, 1490967855104, 4380756541440, 14631844184064],
    "happiness_index": [6.94, 7.16, 6.66, 7.07, 6.38, 6.4, 7.23, 7.22, 5.87, 5.12]
})

# Instantiate a LLM
from pandasai.llm import OpenAI
llm = OpenAI(api_token="YOUR_API_TOKEN")
df = SmartDataframe(df, config={"llm": llm})

'''
New Code
scanner would be used to pass the data to chat option below
'''

scanner = Scanner(config)
df.chat(scanner('Which are the 5 happiest countries?'))
```

**Comment by gventuri** [comment]
Hi @fakabbir thanks a lot for highlighting the 2 points. I'd say let's opt in, as you suggested, for the second option for the time being. We can definitely integrate it at a later point. Or we could use optional dependencies from poetry. What do you think? We've done something similar in other cases (like langchain integration, etc...).

Let me know your thoughts!
